src/flp/types.rs

```diff
@@ -419,55 +417,90 @@ impl<F: FftFriendlyFieldElement> Type for Histogram<F> {
     }
 }
 
-/// A sequence of counters. This type uses a neat trick from [[BBCG+19], Corollary 4.9] to reduce
-/// the proof size to roughly the square root of the input size.
+/// A sequence of integers in range `[0, 2^bits)`. This type uses a neat trick from [[BBCG+19],
+/// Corollary 4.9] to reduce the proof size to roughly the square root of the input size.
 ///
 /// [BBCG+19]: https://eprint.iacr.org/2019/188
 #[derive(Debug, PartialEq, Eq)]
-pub struct CountVec<F, S> {
+pub struct SumVec<F: FftFriendlyFieldElement, S> {
     range_checker: Vec<F>,
     len: usize,
+    bits: usize,
+    flattened_len: usize,
+    max: F::Integer,
     chunk_len: usize,
     gadget_calls: usize,
     phantom: PhantomData<S>,
 }
 
-impl<F: FftFriendlyFieldElement, S: ParallelSumGadget<F, BlindPolyEval<F>>> CountVec<F, S> {
-    /// Returns a new [`CountVec`] with the given length.
-    pub fn new(len: usize) -> Self {
+impl<F: FftFriendlyFieldElement, S: ParallelSumGadget<F, BlindPolyEval<F>>> SumVec<F, S> {
+    /// Returns a new [`SumVec`] with the bit width and vector length.
+    ///
+    /// # Errors
+    ///
+    /// * The length of the encoded measurement, i.e., `bits * len`, overflows addressible memory
+    /// * The bit width cannot be encoded, i.e., `bits` is larger than the number of bits in
+    /// required to encode field elements.
+    ///
+    /// Behavior is undefined if the size of `F::Integer` exceeds `F::ENCODED_SIZE`.
+    pub fn new(bits: usize, len: usize) -> Result<Self, FlpError> {
+        let flattened_len = bits.checked_mul(len).ok_or_else(|| {
+            FlpError::InvalidParameter("`bits*len` overflows addressible memory".into())
+        })?;
+
+        // Check if the bit width might exceed its limit. This limit is defined to be the number of
+        // bits required to encode finite field elements. If the size of `F::Integer` is less than
+        // or equal to `F::ENCODED_SIZE`, then `max` below is guaranteed to not overflow.
```

Comment:
Nit: technically, if `bits` is equal to `F::ENCODED_SIZE * 8`, then the `one << bits` shift will overflow and panic in debug builds. We'd need to use something like `1u128.overflowing_neg().0` with right shifts to safely compute `max` in this case, but that would require more trait jenga.

SumVec with `bits` equal to the field modulus's bit length is of limited use, since any one measurement could wrap around the sum. Thus, I think we should just say not to use it with `bits` equal to the bit width of the field/integer, and update the below error's boundary by one.

Reply:
Good catch, done.

One minor note here: In theory, FLPs might have applications beyond aggregation, i.e., `SumVec` *could* be used for something besides summing. Of course we have constructed our API explicitly around aggregation, so this is maybe a moot point.